github/Auth.py

```diff
@@ -190,36 +193,49 @@ class AppAuth(JWT):
 
     """
 
+    @staticmethod
+    def create_jwt_sign(private_key_or_func: Union[str, PrivateKeyGenerator], jwt_algorithm: str) -> DictSignFunction:
+        def jwt_sign(payload: dict) -> Union[str, bytes]:
+            if callable(private_key_or_func):
+                private_key = private_key_or_func()
+            else:
+                private_key = private_key_or_func
+            return jwt.encode(payload, key=private_key, algorithm=jwt_algorithm)
+
+        return jwt_sign
+
+    # v3: move * above private_key
     def __init__(
         self,
         app_id: Union[int, str],
-        private_key: str,
+        private_key: Optional[Union[str, PrivateKeyGenerator]] = None,
+        *,
+        sign_func: Optional[DictSignFunction] = None,
         jwt_expiry: int = Consts.DEFAULT_JWT_EXPIRY,
         jwt_issued_at: int = Consts.DEFAULT_JWT_ISSUED_AT,
-        jwt_algorithm: str = Consts.DEFAULT_JWT_ALGORITHM,
     ):
         assert isinstance(app_id, (int, str)), app_id
         if isinstance(app_id, str):
             assert len(app_id) > 0, "app_id must not be empty"
-        assert isinstance(private_key, str)
-        assert len(private_key) > 0, "private_key must not be empty"
+        assert private_key is not None or sign_func is not None, "either private_key or sign_func must be given"
+        assert private_key is None or sign_func is None, "private_key or sign_func cannot both be given"
+        if private_key is not None:
+            assert isinstance(private_key, str) or callable(private_key)
+            if isinstance(private_key, str):
+                assert len(private_key) > 0, "private_key must not be empty"
+            sign_func = AppAuth.create_jwt_sign(private_key, Consts.DEFAULT_JWT_ALGORITHM)
         assert isinstance(jwt_expiry, int), jwt_expiry
         assert Consts.MIN_JWT_EXPIRY <= jwt_expiry <= Consts.MAX_JWT_EXPIRY, jwt_expiry
 
         self._app_id = app_id
-        self._private_key = private_key
+        self._sign_func = sign_func
         self._jwt_expiry = jwt_expiry
         self._jwt_issued_at = jwt_issued_at
-        self._jwt_algorithm = jwt_algorithm
 
     @property
     def app_id(self) -> Union[int, str]:
         return self._app_id
 
-    @property
-    def private_key(self) -> str:
```

Comment:
This turned out to be a breaking change, but it's only documented in the features section of the change log…